File: ios/Sources/NearbyMultipeerPlugin/BleCommandQueue.swift
import Foundation

/**
 * Cola de comandos BLE para evitar operaciones concurrentes que pueden causar fallos
 */
class BleCommandQueue {
    private var commandQueue: [() -> Void] = []
    private var isExecuting = false
    private var defaultTimeout: TimeInterval = 10.0 // 10 segundos por defecto
    
    /**
     * Encola un comando para ejecución secuencial con timeout específico
     * - Parameters:
     *   - command: Comando a ejecutar
     *   - timeout: Timeout opcional en segundos
     * - Returns: Resultado del comando
     */
    func enqueue<T>(_ command: @escaping () -> T, timeout: TimeInterval? = nil) -> T {
        let semaphore = DispatchSemaphore(value: 0)
        var result: T!
        var capturedError: Error?
        
        // Encolar el comando
        commandQueue.append {
            do {
                // Ejecutar con timeout
                let timeoutHandler = DispatchWorkItem {
                    capturedError = NSError(domain: "BleCommandQueue", code: -1, userInfo: [
                        NSLocalizedDescriptionKey: "Comando expiró después de \(timeout ?? self.defaultTimeout) segundos"
                    ])
                    semaphore.signal()
                }
                
                // Ejecutar el comando en el hilo principal
                DispatchQueue.main.async {
                    result = command()
                    // Cancelar el timeout
                    timeoutHandler.cancel()
                    semaphore.signal()
                }
                
                // Configurar timeout si está especificado
                if let timeout = timeout, timeout > 0 {
                    DispatchQueue.global().asyncAfter(deadline: .now() + timeout, execute: timeoutHandler)
                } else if self.defaultTimeout > 0 {
                    DispatchQueue.global().asyncAfter(deadline: .now() + self.defaultTimeout, execute: timeoutHandler)
                }
            } catch {
                capturedError = error
                semaphore.signal()
            }
        }
        
        // Iniciar la ejecución si no hay otros comandos en proceso
        if !isExecuting {
            executeNext()
        }
        
        // Esperar a que se complete el comando
        _ = semaphore.wait(timeout: .distantFuture)
        
        // Verificar si hubo error durante la ejecución
        if let error = capturedError {
            print("Error ejecutando comando BLE: \(error.localizedDescription)")
            // No podemos lanzar excepciones dentro de un closure en Swift, así que imprimimos el error
            // y retornamos el valor por defecto para evitar un crash
        }
        
        return result
    }
    
    /**
     * Encola un comando sin retorno
     * - Parameter command: Comando a ejecutar
     */
    func enqueueVoid(_ command: @escaping () -> Void) {
        enqueue {
            command()
            return ()
        }
    }
    
    /**
     * Encola un comando que retorna una promesa
     * - Parameters:
     *   - command: Comando a ejecutar que retorna una promesa
     *   - timeout: Timeout opcional en segundos
     * - Returns: Promesa con el resultado del comando
     */
    func enqueueAsync<T>(_ command: @escaping () -> Promise<T>, timeout: TimeInterval? = nil) -> Promise<T> {
        let promise = Promise<T>()
        
        enqueueVoid {
            let innerPromise = command()
            innerPromise.then { result in
                promise.resolve(result)
            }.catch { error in
                promise.reject(error)
            }
        }
        
        return promise
    }
    
    private func executeNext() {
        guard !commandQueue.isEmpty else {
            isExecuting = false
            return
        }
        
        isExecuting = true
        let command = commandQueue.removeFirst()
        command()
        
        // Ejecutar el siguiente comando
        DispatchQueue.main.async { [weak self] in
            self?.executeNext()
        }
    }
    
    /**
     * Establece el timeout por defecto para todos los comandos
     * - Parameter timeout: Timeout en segundos (0 para desactivar)
     */
    func setDefaultTimeout(_ timeout: TimeInterval) {
        defaultTimeout = timeout
    }
}

/**
 * Clase simple de Promise para manejar operaciones asíncronas
 */
class Promise<T> {
    typealias ResolveCallback = (T) -> Void
    typealias RejectCallback = (Error) -> Void
    
    private var resolveCallbacks: [ResolveCallback] = []
    private var rejectCallbacks: [RejectCallback] = []
    private var value: T?
    private var error: Error?
    private var isResolved = false
    private var isRejected = false
    
    func resolve(_ value: T) {
        guard !isResolved && !isRejected else { return }
        
        self.value = value
        isResolved = true
        
        resolveCallbacks.forEach { $0(value) }
        resolveCallbacks.removeAll()
        rejectCallbacks.removeAll()
    }
    
    func reject(_ error: Error) {
        guard !isResolved && !isRejected else { return }
        
        self.error = error
        isRejected = true
        
        rejectCallbacks.forEach { $0(error) }
        resolveCallbacks.removeAll()
        rejectCallbacks.removeAll()
    }
    
    @discardableResult
    func then(_ callback: @escaping ResolveCallback) -> Promise<T> {
        if isResolved, let value = value {
            callback(value)
        } else {
            resolveCallbacks.append(callback)
        }
        
        return self
    }
    
    @discardableResult
    func `catch`(_ callback: @escaping RejectCallback) -> Promise<T> {
        if isRejected, let error = error {
            callback(error)
        } else {
            rejectCallbacks.append(callback)
        }
        
        return self
    }
} 

File: android/src/main/java/com/squareetlabs/capacitor/nearbymultipeer/BleCommandQueue.java
package com.squareetlabs.capacitor.nearbymultipeer;

import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import java.util.LinkedList;
import java.util.Queue;
import java.util.concurrent.Callable;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

/**
 * Cola de comandos BLE para evitar operaciones concurrentes que pueden causar fallos
 */
public class BleCommandQueue {
    private static final String TAG = "BleCommandQueue";
    private final Queue<QueueItem<?>> commandQueue = new LinkedList<>();
    private boolean isExecuting = false;
    private final Handler handler = new Handler(Looper.getMainLooper());
    private long defaultTimeout = 10000; // 10 segundos por defecto

    /**
     * Encola un comando para ejecución secuencial
     * @param command Comando a ejecutar
     * @param <T> Tipo de retorno del comando
     * @return Resultado del comando
     */
    public synchronized <T> T enqueue(Callable<T> command) throws Exception {
        return enqueue(command, defaultTimeout);
    }

    /**
     * Encola un comando para ejecución secuencial con timeout específico
     * @param command Comando a ejecutar
     * @param timeout Timeout en milisegundos
     * @param <T> Tipo de retorno del comando
     * @return Resultado del comando
     */
    public synchronized <T> T enqueue(Callable<T> command, long timeout) throws Exception {
        QueueItem<T> queueItem = new QueueItem<>(command);
        commandQueue.add(queueItem);
        
        if (!isExecuting) {
            executeNext();
        }
        
        // Esperar a que se complete el comando o ocurra un timeout
        if (timeout > 0) {
            long startTime = System.currentTimeMillis();
            while (!queueItem.isCompleted() && System.currentTimeMillis() - startTime < timeout) {
                try {
                    TimeUnit.MILLISECONDS.sleep(10);
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    throw new Exception("Command interrupted", e);
                }
            }
            
            if (!queueItem.isCompleted()) {
                throw new TimeoutException("Command timed out after " + timeout + "ms");
            }
        } else {
            // Esperar indefinidamente
            while (!queueItem.isCompleted()) {
                try {
                    TimeUnit.MILLISECONDS.sleep(10);
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    throw new Exception("Command interrupted", e);
                }
            }
        }
        
        if (queueItem.getException() != null) {
            throw queueItem.getException();
        }
        
        return queueItem.getResult();
    }
    
    /**
     * Encola un comando sin retorno (Runnable)
     * @param command Comando a ejecutar
     */
    public synchronized void enqueueRunnable(Runnable command) throws Exception {
        enqueue(() -> {
            command.run();
            return null;
        });
    }

    private synchronized void executeNext() {
        if (commandQueue.isEmpty()) {
            isExecuting = false;
            return;
        }
        
        isExecuting = true;
        QueueItem<?> queueItem = commandQueue.poll();
        
        handler.post(() -> {
            try {
                Object result = queueItem.getCommand().call();
                queueItem.setResult(result);
            } catch (Exception e) {
                Log.e(TAG, "Error executing command", e);
                queueItem.setException(e);
            } finally {
                queueItem.setCompleted(true);
                executeNext();
            }
        });
    }
    
    /**
     * Establece el timeout por defecto para todos los comandos
     * @param defaultTimeout Timeout en milisegundos (0 para desactivar)
     */
    public void setDefaultTimeout(long defaultTimeout) {
        this.defaultTimeout = defaultTimeout;
    }
    
    /**
     * Clase para almacenar un comando en la cola con su resultado/excepción
     */
    private static class QueueItem<T> {
        private final Callable<T> command;
        private T result;
        private Exception exception;
        private boolean completed = false;
        
        public QueueItem(Callable<T> command) {
            this.command = command;
        }
        
        public Callable<T> getCommand() {
            return command;
        }
        
        public T getResult() {
            return result;
        }
        
        public void setResult(Object result) {
            this.result = (T) result;
        }
        
        public Exception getException() {
            return exception;
        }
        
        public void setException(Exception exception) {
            this.exception = exception;
        }
        
        public boolean isCompleted() {
            return completed;
        }
        
        public void setCompleted(boolean completed) {
            this.completed = completed;
        }
    }
} 